### PR TITLE
Performance optimizations

### PR DESCRIPTION
**Pull request by workofart**
## Description
* Updated the performance test to be more fine-grain to capture start-to-peak  metrics, not just peak metrics, to avoid being influenced higher start up "baselines". Also, updated summary metrics that's less biased and sensitive to system fluctuations
* Optimize host syncs and extra Python work in the hot path
  * Removes `.item()` and other scalar extractions from the training hot path.
  * `train_step()` now returns a Tensor; loss accumulation uses `loss.data`.
  * Grad clipping avoids `.item()` inside loops.
* Embedding.forward uses raw index arrays directly
  * Avoids wrapping index arrays into `Tensor` only to immediately access `.data`.
  * Applies to token + position embeddings in GPT forward.
  * Removes minor redundant object construction; negligible relative to full training step.
* Avoid retaining function + inputs for constant-only ops
  * `Function.apply()` skips attaching autograd creator when no inputs require grad.
  * Reduces unnecessary graph nodes and Python object churn.
* Introduce `no_grad()` context manager
  * Adds ability to explicitly disable gradient tracking. This is different from the existing `model.eval()` that changes the BatchNorm and Dropout behavior.
* Preserve explicitly constructed backend array dtypes
  * Stops forcing backend arrays to `float32`; keeps original dtype (`int32`, `float16`, etc.).
  * Improves correctness/control over dtype handling.
* add throughput metrics in dataloader/collator and display in trainer tqdm progress bar and clean up dataloader contracts
  ```
  Training Batches: 25it [00:00, 48.72it/s, tokens=19,200, tok/s=27,719.9]
  ```

## Tests

Added unit tests.

## Performance Comparison

Even though some of the optimizations didn't show significant improvements, they're are in-theory worthwhile, so I'm still implementing them.

**Final**

| Case              | Step P50 | Step P90 | GFLOP/s       | Graph Node Counts | Static Memory | Memory Peak Over Start |
|-------------------|----------|----------|---------------|-------------|---------------|------------------------|
| Complex MLP Model | 2.35 ms  | 2.46 ms  | 3,458 GFLOP/s | 39          | 40.08 KiB     | 4.56 GiB               |
| Deep CNN Model    | 19.38 ms | 20.28 ms | 520.4 GFLOP/s | 115         | 5.79 MiB      | 1.75 GiB               |
| Mini GPT-2 Model  | 856.1 ms | 861.2 ms | 4,966 GFLOP/s | 413         | 777.94 MiB    | 10.81 GiB              |

**Baseline**

| Case              | Step P50 | Step P90 | GFLOP/s       | Graph Node Counts | Static Memory | Memory Peak Over Start |
|-------------------|----------|----------|---------------|-------------|---------------|------------------------|
| Complex MLP Model | 2.35 ms  | 2.44 ms  | 3,450 GFLOP/s | 39          | 40.08 KiB     | 4.71 GiB               |
| Deep CNN Model    | 19.35 ms | 22.62 ms | 521.1 GFLOP/s | 115         | 5.79 MiB      | 1.74 GiB               |
| Mini GPT-2 Model  | 873 ms   | 873.4 ms | 4,870 GFLOP/s | 413         | 777.94 MiB    | 11.63 GiB              |

**Delta Improvements**

| Commit | Mini GPT-2 ΔStep P50 | Mini GPT-2 ΔStep P90 | Mini GPT-2 ΔGFLOP/s | Mini GPT-2 ΔNode Counts | Mini GPT-2 ΔStatic Memory | Mini GPT-2 ΔMemory Peak Over Start |
|---|---|---|---|---|---|---|
| Baseline | +0.00 ms | +0.00 ms | +0 | +0 | +0.00 MiB | +0.00 GiB |
| Optimize host syncs and extra Python work in the hot path | -41.30 ms | -38.50 ms | +223 | +0 | +0.00 MiB | +0.79 GiB |
| Embedding.forward now uses raw index arrays directly | +2.30 ms | +0.80 ms | -13 | +0 | +0.00 MiB | +0.00 GiB |
| avoid retaining the function and its inputs for constant-only ops | -1.90 ms | -1.60 ms | +11 | +0 | +0.00 MiB | +0.00 GiB |
| introduce no_grad() context manager for inference and eval | +1.90 ms | -0.50 ms | -11 | +0 | +0.00 MiB | +0.00 GiB |
| Preserve explicitly constructed backend array dtypes | +6.70 ms | +9.80 ms | -38 | +0 | +0.00 MiB | -1.61 GiB |

### CODE REVIEW

File: autograd/text/tokenizer.py [hunk 1]
```diff
@@ -206,7 +206,7 @@ def prepare_data(
 
         logger.info(f"Vocabulary size: {len(self._unicode_to_int_vocab)}")
         logger.info(f"Encoded data length: {len(encoded_data)}")
-        logger.debug(f"Sample encoded data (first 50 tokens): {encoded_data[:50]}")
+        logger.debug("Sample encoded data (first 50 tokens): %s", encoded_data[:50])
```
Comment (workofart) [inline autograd/text/tokenizer.py:209]: f-formatted strings are actually copying the underlying data to CPU if they're Tensors previously on GPU, regardless if they're printed in the console or not. In this case, encoded_data is not a Tensor, but there are some cases below where they're Tensors, and I'm just doing a general sweep here.

File: autograd/tools/trainer.py [hunk 2]
```diff
@@ -204,16 +209,14 @@ def _save_metrics(self):
...
-        Subclasses must implement the forward pass, loss computation, and backward pass here.
+    def train_step(self, batch_data, data_loader=None) -> Tensor:
+        """Perform a single training step and return the device loss tensor.
 
         Args:
             batch_data: A single batch of training data.
 
         Returns:
-            float: The computed loss value for the batch.
+            Tensor: The computed loss tensor for the batch.
```
Comment (workofart) [inline autograd/tools/trainer.py:249]: Keeping each step as Tensor prevents the copy from GPU to CPU after each step.

File: autograd/tools/trainer.py [hunk 3]
```diff
@@ -446,53 +449,57 @@ def evaluate(self, val_data_loader) -> float:
...
-                if sample_count > 0:
-                    sample_lines = [
-                        f"Predicted: {y_pred_proc[i]}\nActual: {y_true_proc[i]}"
-                        for i in range(sample_count)
-                    ]
-                    logger.info(
-                        "Sample Predictions on Validation Batch:\n"
-                        + "\n".join(sample_lines)
+        try:
+            with no_grad():
```
Comment (workofart) [inline autograd/tools/trainer.py:485]: This is just wrapping with no_grad() to avoid building the computational graph, which is overhead.

File: autograd/nn.py [hunk 4]
```diff
@@ -1176,13 +1179,13 @@ def forward(self, x: Union[Tensor, ArrayLike]) -> Tensor:
             >>> x = Tensor(xp.array([[0, 1, 2], [3, 4, 5]]))
             >>> y = embed(x)  # Expected: (2, 3, 8)
         """
-        if not isinstance(x, Tensor):
```
Comment (workofart) [inline autograd/nn.py:1179]: Avoid building a computational node for a non-Tensor object.

File: autograd/optim.py [hunk 5]
```diff
@@ -253,23 +253,19 @@ def _clip_grad_norm(self, max_norm: float, norm_type: float = 2.0) -> None:
         """
         # Compute the global norm of all gradients
         total_norm = 0.0
-
         for param in self.model_parameters.values():
             if param.grad is not None:
                 grad_data = param.grad.data
-                total_norm += (xp.abs(grad_data) ** norm_type).sum().item()
+                total_norm += (xp.abs(grad_data) ** norm_type).sum()
```
Comment (workofart) [inline autograd/optim.py:259]: Avoid copying to CPU

File: autograd/optim.py [hunk 6]
```diff
@@ -253,23 +253,19 @@ def _clip_grad_norm(self, max_norm: float, norm_type: float = 2.0) -> None:
...
 
         # If total_norm is greater than max_norm, scale all gradients
-        if total_norm > max_norm:
-            scale_factor = max_norm / (
-                total_norm + 1e-10
-            )  # add small value for numerical stability
-            for param in self.model_parameters.values():
-                if param.grad is not None:
-                    param.grad.data *= scale_factor
+        scale_factor = xp.minimum(1.0, max_norm / (total_norm + 1e-10))
```
Comment (workofart) [inline autograd/optim.py:268]: minor logic simplification

File: autograd/tensor.py [hunk 7]
```diff
@@ -423,7 +456,7 @@ def __iadd__(self, other: Union["Tensor", float, int]) -> "Tensor":
             Tensor: This tensor, after in-place addition.
         """
         if not isinstance(other, Tensor):
-            other = Tensor(other)
+            other = Tensor(other, requires_grad=False)
```
Comment (workofart) [inline autograd/tensor.py:465]: avoid creating a node in the computation graph for non-Tensor inputs like (`+= 1` operations, the `1` is the non-Tensor input)

File: autograd/tensor.py [hunk 8]
```diff
@@ -398,6 +427,10 @@ def __pow__(self, other: Union["Tensor", float, int]) -> "Tensor":
         if isinstance(other, int) and other >= 0:
             if other == 0:
                 return Tensor(xp.ones_like(self.data), requires_grad=False)
+            if other == 1:
```
Comment (workofart) [inline autograd/tensor.py:436]: Currently this is bypassing the `no_grad()` context:
```
with no_grad():
    y = x ** 1

assert y is x
assert y.requires_grad is True  # which is wrong
```

File: autograd/nn.py [hunk 9]
```diff
@@ -608,13 +611,17 @@ def __init__(
                         self.in_channels,
                         self.kernel_size,
                         self.kernel_size,
-                    )
+                    ),
+                    dtype=xp.float32,
```
Comment (workofart) [inline autograd/nn.py:615]: For these dtype changes, we're trying to preserve the dtype of individual Tensors, yet still default the `nn.Module` to fp32 for the ones that didn't specify at the Tensor level.

 1. So explicit backend array dtype is still preserved now
```python
# before
x = np.array([1.0, 2.0], dtype=np.float64)
Tensor(x).data.dtype
# float32

# after
x = np.array([1.0, 2.0], dtype=np.float64)
Tensor(x).data.dtype
# float64
```


2. Library-created default parameter


```python
# before
Linear(4, 2)._parameters["weight"].data.dtype
# float32

# after
Linear(4, 2)._parameters["weight"].data.dtype
# float32
```
